File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/interactions/KafkaConfig.java
package pl.allegro.experiments.chi.chiserver.infrastructure.interactions;

import io.micrometer.core.instrument.MeterRegistry;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.Metric;
import org.apache.kafka.common.serialization.ByteArraySerializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.core.DefaultKafkaProducerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import pl.allegro.experiments.chi.chiserver.domain.experiments.EventDefinitionRepository;
import pl.allegro.experiments.chi.chiserver.domain.experiments.EventDefinitionSaver;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;
import pl.allegro.experiments.chi.chiserver.domain.interactions.InteractionRepository;
import pl.allegro.experiments.chi.chiserver.infrastructure.EventDefinitionSaveScheduler;
import pl.allegro.experiments.chi.chiserver.infrastructure.experiments.KafkaEventDefinitionRepository;
import pl.allegro.experiments.chi.chiserver.infrastructure.experiments.LoggerEventDefinitionRepository;
import pl.allegro.tech.common.andamio.server.cloud.CloudMetadata;
import pl.allegro.tech.common.andamio.avro.AvroConverter;

import java.util.*;

@Configuration
public class KafkaConfig {
    private final List<String> metricsList = Arrays.asList("request-rate", "record-send-rate", "request-latency-avg", "request-latency-max", "record-error-rate");

    @Bean
    @ConditionalOnProperty(name = {"interactions.repository"}, havingValue = "local")
    public InteractionRepository localInteractionRepository() {
        return new LoggerInteractionRepository();
    }

    @Bean
    @ConditionalOnProperty(name = {"eventDefinitions.repository"}, havingValue = "local")
    public EventDefinitionRepository localEventDefinitionsRepository() {
        return new LoggerEventDefinitionRepository();
    }

    @Bean
    @ConditionalOnProperty(name = {"interactions.repository"}, havingValue = "kafka")
    public InteractionRepository kafkaInteractionRepository(
            KafkaTemplate<String, byte[]> kafkaTemplate,
            AvroConverter avroConverter,
            @Value("${interactions.kafka.topic}") String kafkaTopic) {
        return new KafkaInteractionRepository(kafkaTemplate, avroConverter, kafkaTopic);
    }

    @Bean
    @ConditionalOnProperty(name = {"eventDefinitions.repository"}, havingValue = "kafka")
    public EventDefinitionRepository kafkaEventDefinitionRepository(
            KafkaTemplate<String, byte[]> kafkaTemplate,
            AvroConverter avroConverter,
            @Value("${eventDefinitions.kafka.topic}") String kafkaTopic) {
        return new KafkaEventDefinitionRepository(kafkaTemplate, avroConverter, kafkaTopic);
    }

    @Bean
    public EventDefinitionSaver eventDefinitionSaver(
            ExperimentsRepository experimentsRepository,
            EventDefinitionRepository eventDefinitionRepository) {
        return new EventDefinitionSaver(eventDefinitionRepository, experimentsRepository);
    }

    @Bean
    @ConditionalOnProperty(name = {"kafka.template"}, havingValue = "on")
    public KafkaTemplate kafkaTemplate(
            @Value("${kafka.template}") String kafka,
            @Value("${kafka.bootstrap-servers-dc4}") String bootstrapServersDc4,
            @Value("${kafka.bootstrap-servers-dc5}") String bootstrapServersDc5,
            CloudMetadata cloudMetadata,
            @Value("${kafka.batch-size}") int batchSize,
            @Value("${kafka.linger-ms}") int lingerMs,
            MeterRegistry metricRegistry) {
        String bootstrapServer = cloudMetadata.getDatacenter().equals("dc5") ? bootstrapServersDc5 : bootstrapServersDc4;
        final KafkaTemplate<String, byte[]> kafkaTemplate = new KafkaTemplate<>(producerFactory(bootstrapServer, batchSize, lingerMs));

        metricsList.forEach(metricName ->
                metricRegistry.gauge("kafka." + metricName, metricName,
                        mName -> getMetricValue(kafkaTemplate, metricName ) )
        );

        return kafkaTemplate;
    }

    @Bean
    EventDefinitionSaveScheduler eventDefinitionSaveScheduler(EventDefinitionSaver eventDefinitionSaver) {
        return new EventDefinitionSaveScheduler(eventDefinitionSaver);
    }

    private double getMetricValue(KafkaTemplate<String, byte[]> kafkaTemplate, String metricName) {
        return kafkaTemplate.metrics().keySet().stream()
                .filter(it -> it.name().equals(metricName) && it.tags().size() < 2)
                .findFirst()
                .flatMap(it -> {
                    Metric metric = kafkaTemplate.metrics().get(it);
                    return Optional.ofNullable(metric != null ? metric.value() : null);
                })
                .orElse(.0);
    }

    private DefaultKafkaProducerFactory<String, byte[]> producerFactory(
            String bootstrapServers,
            int batchSize,
            int lingerMs) {
        return new DefaultKafkaProducerFactory<>(config(bootstrapServers, batchSize, lingerMs));
    }

    private Map<String, Object> config(
            String bootstrapServers,
            int batchSize,
            int lingerMs) {
        Map<String, Object> result = new HashMap<>();
        result.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        result.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
        result.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        result.put(ProducerConfig.BATCH_SIZE_CONFIG, batchSize);
        result.put(ProducerConfig.LINGER_MS_CONFIG, lingerMs);
        result.put(ProducerConfig.ACKS_CONFIG, "1");
        result.put(ProducerConfig.RETRIES_CONFIG, "5");
        result.put(ProducerConfig.RETRY_BACKOFF_MS_CONFIG, "1000");
        return result;
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/interactions/infrastructure/KafkaConfig.kt
package pl.allegro.experiments.chi.chiserver.interactions.infrastructure

import com.codahale.metrics.Gauge
import com.codahale.metrics.MetricRegistry
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.MetricName
import org.apache.kafka.common.serialization.ByteArraySerializer
import org.apache.kafka.common.serialization.StringSerializer
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.core.DefaultKafkaProducerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.kafka.core.ProducerFactory
import pl.allegro.experiments.chi.chiserver.interactions.InteractionRepository
import pl.allegro.tech.common.andamio.server.cloud.CloudMetadata
import pl.allegro.tech.common.andamio.spring.avro.AvroConverter
import java.util.*

@Configuration
class KafkaConfig {
    private val metricsList = listOf("request-rate", "record-send-rate", "request-latency-avg", "request-latency-max", "record-error-rate")

    @Bean
    @ConditionalOnProperty(name = arrayOf("interactions.repository"), havingValue = "local")
    fun localInteractionRepository(): InteractionRepository {
        return LoggerInteractionRepository()
    }

    @Bean
    @ConditionalOnProperty(name = arrayOf("interactions.repository"), havingValue = "kafka")
    fun kafkaInteractionRepository(
            kafkaTemplate: KafkaTemplate<String, ByteArray>,
            avroConverter: AvroConverter,
            @Value("\${interactions.kafka.topic}") kafkaTopic: String): InteractionRepository {
        return KafkaInteractionRepository(kafkaTemplate, avroConverter, kafkaTopic)
    }

    @Bean
    @ConditionalOnProperty(name = arrayOf("interactions.repository"), havingValue = "kafka")
    fun kafkaTemplate(
            @Value("\${interactions.kafka.bootstrap-servers-dc4}") bootstrapServersDc4: String,
            @Value("\${interactions.kafka.bootstrap-servers-dc5}") bootstrapServersDc5: String,
            cloudMetadata: CloudMetadata,
            @Value("\${interactions.kafka.batch-size}") batchSize: Int,
            @Value("\${interactions.kafka.linger-ms}") lingerMs: Int,
            metricRegistry : MetricRegistry
    ): KafkaTemplate<String, ByteArray> {

        val bootstrapServer = if (cloudMetadata.datacenter == "dc5") {
            bootstrapServersDc5
        } else { // dc4 or localhost
            bootstrapServersDc4
        }

        val kafkaTemplate =  KafkaTemplate<String, ByteArray>(producerFactory(bootstrapServer, batchSize, lingerMs))

        metricsList.forEach { metricName: String ->
            metricRegistry.register("kafka." + metricName, Gauge<Double> {
                getMetricValue(kafkaTemplate, metricName)
            })
        }

        return kafkaTemplate
    }

    private fun getMetricValue(kafkaTemplate : KafkaTemplate<String, ByteArray>, metricName : String) : Double {
        return kafkaTemplate.metrics().keys.stream()
                .filter {it.name() == metricName && it.tags().size < 2}.findFirst()
                .flatMap {Optional.ofNullable(kafkaTemplate.metrics()[it]?.value())}
                .orElse(.0)
    }

    private fun producerFactory(
            bootstrapServers: String,
            batchSize: Int,
            lingerMs: Int): ProducerFactory<String, ByteArray> =
            DefaultKafkaProducerFactory<String, ByteArray>(config(bootstrapServers, batchSize, lingerMs))

    private fun config(
            bootstrapServers: String,
            batchSize: Int,
            lingerMs: Int) =
            mapOf(
                    ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to StringSerializer::class.java,
                    ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to ByteArraySerializer::class.java,
                    ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to bootstrapServers,
                    ProducerConfig.BATCH_SIZE_CONFIG to batchSize,
                    ProducerConfig.LINGER_MS_CONFIG to lingerMs,
                    ProducerConfig.ACKS_CONFIG to "1",
                    ProducerConfig.RETRIES_CONFIG to "5",
                    ProducerConfig.RETRY_BACKOFF_MS_CONFIG to "1000"
            )
}
